cmd/device-worker/main.go

```diff
@@ -147,9 +151,36 @@ func main() {
 
 	setupSignalHandler(metricsStore)
 
+	go ListenStartGracefulRebootChannel(err, wl, dataMonitor, hbs, deviceOs)
+
 	if err := s.Serve(l); err != nil {
 		log.Fatalf("cannot start worker server, err: %v", err)
 	}
+
+}
+
+func ListenStartGracefulRebootChannel(err error, wl *workload2.WorkloadManager, dataMonitor *datatransfer.Monitor, hbs *heartbeat2.Heartbeat, deviceOs *os2.OS) {
+	// listen to the channel for getting StartGracefulReboot signal
+	for {
+		select {
+		case job := <-GracefulRebootChannel:
+			if job == os2.StartGracefulReboot {
+				log.Info("A graceful reboot request was received")
+				if err = wl.GracefulReboot(); err != nil {
+					log.Fatalf("cannot graceful reboot the workloads: %v", err)
+				}
+
+				if err = dataMonitor.Deregister(); err != nil {
```

Comment:
I think it is worth deregistering `systemMetricsWatcher` and `metricsStore` - in that order.

Reply:
In the systemMetricsWatcher's Deregistration should be something more than deleting the target "system"?

Reply:
No, that's it. 
It needs to be done before `metricsStore` or else there might be errors (from writing to a closed DB) in the logs.